### ISSUE TITLE
[Feature-Request] Allow adding of entries by just specifying label & secret

### ISSUE THREAD

**Issue by ChrLau** [freeotp/freeotp-android#360]
Hi,

I set up 2FA on my LinkedIn account today and ran into some problem with FreeOTP.
Root cause is that LinkedIn doesn't display a QR-Code anymore (got removed when they where bought by Microsoft it seems). They just display the secret key and tell you to enter that into the Microsoft Authenticator App.

Google Authenticator seems to have such a feature too.

But FreeOTP lacks this. And as LinkedIn was providing no technical information about the algorithm and the like I wasn't able to make it work initially. LinkedIn reported that the entered code wasn't valid.
 
What worked however was to build the most basic otpauth:// string myself and generate a QR-Code from that.
`otpauth://totp/LinkedIn:MyAccount?secret=BCUTUKT6MT33Y3FCHSMHU2XRHKLTX7RN`

After scanning this QR-Code with FreeOTP the code was accepted by LinkedIn.

Is it possible to add such a feature to FreeOTP?
Some dialogue where I can only choose if it's TOTP or HOTP, specify the label, the secret and that's it? Omit all other parameters. If you need that, you can still use the already existing dialogue.

Would save me a few minutes even with termux and qrencode installed on my Android phone. ;-)

**Comment by justin-stephenson** [comment]
FreeOTP includes this 'Manual add' feature already. When you press the `+` Icon to add a token, press the `pencil` button instead of the QR code button. Can you please check if this works for you?

**Comment by ChrLau** [comment]
Hi, I tried using that feature. But I think the problem is that there are always the parameters for algorithm and interval included. And there is no way to make them optional.

**Comment by justin-stephenson** [comment]
If algorithm and period are not provided as part of the otpauth [URI](https://github.com/google/google-authenticator/wiki/Key-Uri-Format), then FreeOTP assumes `SHA1` and `30`, does the code fail when these values are selected ?

**Comment by ChrLau** [comment]
Deactivated 2FA on LinkedIn and set it up again. Choosing SHA1 and interval auf 30 works. I assumed SHA256 is the default is this is always pre-selected.

**Comment by justin-stephenson** [comment]
Are you okay to close this ticket or do you see some improvement needed still in FreeOTP?

**Comment by justin-stephenson** [comment]
I agree. We should improve the manual token UI similar to https://stefansundin.github.io/2fa-qr/ which allows removing and adding the optional otpauth parameters such as algorithm, digits, and period.

**Comment by ChrLau** [comment]
That would be the best way IMHO.

**Comment by rockdaboot** [comment]
I ran into the issue today and it still is quite a nuisance.
Setting up LinkedIn 2FA failed for me on the first try. Thanks to the well written [instructions](https://admin.brennt.net/howto-use-freeotp-for-two-factor-authentication-2fa-on-linkedin) from @ChrLau my second try succeeded.

**Comment by ChrLau** [comment]
> I ran into the issue today and it still is quite a nuisance. Setting up LinkedIn 2FA failed for me on the first try. Thanks to the well written [instructions](https://admin.brennt.net/howto-use-freeotp-for-two-factor-authentication-2fa-on-linkedin) from @ChrLau my second try succeeded.

Thanks for the flowers. 😄 Reminded me to add a TL;DR to my post. As the whole "generate a QR code" part can be omitted if you figured out the parameters LinkedIn requires.

**Comment by Ditto89** [comment]
Hello!
I'm having the same problem with my organization, which uses the Microsoft login to authenticate on everything (Outlook and stuff). From this week on, the FreeOTP keys I used to use won't work anymore, so I had to change to the MS Auth App... =/
Apparently, they removed/blocked access to the option to add any auth app as 2FA, and, when trying to read the only QR available, FreeOTP gives me "invalid code".
I tried to follow @ChrLau [instuctions](https://admin.brennt.net/howto-use-freeotp-for-two-factor-authentication-2fa-on-linkedin) and used the parameters to LinkedIn from the "TL;DR" section to add manually, but they failed.
Do you know which settings to use at the Microsoft login form, specifically? I don't think that "otpauth://totp/**LinkedIn**:MyAccount?secret=KEY-YOU-GOT-FROM-LINKEDIN" would work for my case, since they are different websites... (and I'm on Windows, by the way).
Thank you very much!

**Comment by ChrLau** [comment]
Hi @Ditto89 

1. That LinkedIn is just a text used to display the description. It can be changed freely.
2. Sadly you can't replace MS Authenticator with FreeOTP in your case. While MS Authenticator can use the standard HOTP/TOTP method where a new key/pin is generator every X seconds, for Outlook, Teams etc. this is different. Here Microsoft uses a so-called "Out-of-Band (OOB) Push Authentication" method. This is where you get a push notification to your mobile and are asked to enter the number displayed an your computer and acknowledge that. This is not standarized completely and FreeOTP doesn't support it.
3. Additionally your company's IT people can configure that only the MS Authenticator App is allowed to sign-in. This is also true in my case. Hence I have to use the MS app..

**Comment by Ditto89** [comment]
Hello @ChrLau and everyone!

Thanks for your reply!

Luckily, I'm still using OTPs generated by the MS Auth app (instead of push notifications), so, _theoretically_, they still could be added to FreeOTP, too (I'm hoping they didn't block other apps besides MS Auth, though...).

To try that, I've generated a QR Code using "otpauth://totp/MS:MyAccount?secret=?????", but I don't know exactly what to put on that "?????". I've tried using the codes MS shows when adding a new 2FA method (as explained below), both via QR Code and manually on FreeOTP (Algorithm: SHA1, Interval: 30 seconds, etc.) but they didn't work.

Sooo... Could you, please, point me up what parameter(s) goes in that "?????" part of the QR Code (or the parameters for manual add) ?

Just to recall: in the MS Account Settings, under "Security Info", to add a 2FA app, they give me a QR Code to scan, a 9-digits numeric code, and an URL that looks like "mobileappcommunicator.auth.microsoft.com/activatev2 / ANOTHER-9-DIGITS-CODE / 18-CHARS-ALPHANUMERIC-CODE
(I hope I'm making myself clear...)

Thank you very much!

**Comment by ChrLau** [comment]
Hi, sorry but that doesn't tell me much and it looks like mobileappcommunicator.auth.microsoft.com is tied to the proprietary method (the one with push notifications to your mobile where the MS app pops up and asks you to enter a 2 digit number you are shown).
In my private Microsoft Account I have the option to say to use another app for authentication. Instead of the MS Authenticator. This is what you need to click. Then you will get a QR-Code and this will have the following information in it:

otpauth://totp/Microsoft:MAIL@DOMAIN?secret=**randomNumbersAndLetters**&issuer=Microsoft

You would need to enter the bold part as secret. This is the base32 encoded cryptographic key which is needed for your 2FA app to generate the same number tokens as the authentication server you are authenticating against.

Just to make it clear, the secret is nothing I can give you or calculate in any way. As it is generated by the authentication server.

**EDIT November 2025:** As I regularly get questions about this, here is a link to the Microsoft documentation: https://learn.microsoft.com/en-us/entra/identity/authentication/how-to-mfa-number-match and apparently the method where a user has to enter a 2 digit long number is called "Number matching".

**Comment by Ditto89** [comment]
Hello!

Thanks for your help! And sorry for kind of messing up your thread... =/

Well, I tried some parameters on FreeOTP manual add, and I even generated QRCodes using that "otpauth://xxxxx" parameters, but it seems I really lack that secret key to get the corrects OTPs on FreeOTP...

Regarding that, have some of you tried this, for example?
https://fekir.info/post/extract-tokens-from-microsoft-authenticator/
https://sarange.gitlab.io/posts/extracting-totp-from-microsoft-authenticator/

My phone is not rooted, so I could not try, but I put it here just as a suggestion to those who want to try to retrieve their own secret keys from the MS Auth app, and get rid of it... I hope it helps someway.

Thanks again!
